syft/frameworks/torch/mpc/primitives.py

```diff
@@ -0,0 +1,169 @@
+        # tensorized key generation algorithms.
+        self.fss_eq: list = []
+        self.fss_comp: list = []
+        self.beaver: list = []
+        self.xor_add_couple: list = []  # couple of the same value shared via ^ or + op
+
+        self._owner: AbstractWorker = owner
+        self._builders: dict = {
+            "fss_eq": self.build_fss_keys(type_op="eq"),
+            "fss_comp": self.build_fss_keys(type_op="comp"),
+            "beaver": self.build_triples,
+            "xor_add_couple": self.build_xor_add_couple,
+        }
+
+    def get_keys(self, type_op, n_instances=1, remove=True):
+        """
+        Return FSS keys primitives #TODO
+
+        Args:
+            type_op: fss_eq, fss_comp, or xor_add_couple
+            n_instances: how many primitives to retrieve. Comparison is pointwise so this in
+                convenient
+            remove: if true, pop out the primitive. If false, only read it. Read mode is
+                needed because we're working on virtual workers and they need to gather
+                a some point and then re-access the keys.
+        """
+        primitive_stack = getattr(self, type_op)
+
+        available_instances = len(primitive_stack[0]) if len(primitive_stack) > 0 else -1
+        if available_instances >= n_instances:
+            keys = []
+            # We iterate on the different elements that constitute a given primitive, for
+            # example of the beaver triples, you would have 3 elements.
+            for i, prim in enumerate(primitive_stack):
+                # We're selecting on the last dimension of the tensor because it's simpler for
+                # generating those primitives in crypto protocols
+                # [:] ~ [slice(None)]
+                # [:1] ~ [slice(1)]
+                # [1:] ~ [slice(1, None)]
+                # [:, :, :1] ~ [slice(None)] * 2 + [slice(1)]
+                n_dim = len(prim.shape)
+                get_slice = [slice(None)] * (n_dim - 1) + [slice(n_instances)]
+                remaining_slice = [slice(None)] * (n_dim - 1) + [slice(n_instances, None)]
+
+                keys.append(prim[get_slice])
+                if remove:
+                    primitive_stack[i] = prim[remaining_slice]
+
+            return keys
+        else:
+            raise EmptyCryptoPrimitiveStoreError(self, type_op, available_instances, n_instances)
+
+    def provide_primitives(
+        self,
+        crypto_types: Union[str, List[str]],
+        workers: List[AbstractWorker],
+        n_instances: int = 10,
+        **kwargs,
+    ):
+        """
+        Build n_instances of crypto primitives of the different crypto_types given and send them to some workers.
+
+        Args:
+            crypto_types: type of primitive (fss_eq, etc)
+            workers: recipients for those primitive
+            n_instances: how many of them are needed
+            **kwargs: any parameters needs for the primitive builder
+
+        Returns:
+
+        """
```

Comment:
Not sure I followed everything but this method never uses already created primitives, does it?
Do you think it could call `get_keys()` and `get_keys()` would return or create keys if there are not enough of them?

Reply:
This method is here to feed the different crypto stores, so shat get-keys will work.
It's here to decorrelate preprocessing and online phase: provide_primitives should be called way before the computation really happens, to have a fats online phase